Sources/Valet/Internal/Keychain.swift

```diff
@@ -48,216 +48,173 @@ internal final class Keychain {
+            try SecItem.add(attributes: secItemQuery)
         #else
-            
-            if case .success = containsObject(forKey: key, options: options) {
-                return SecItem.update(attributes: [kSecValueData as String: object], forItemsMatching: secItemQuery)
-            } else {
-                secItemQuery[kSecValueData as String] = object
-                return SecItem.add(attributes: secItemQuery)
-            }
+
+        if containsObject(forKey: key, options: options) == errSecSuccess {
+            try SecItem.update(attributes: [kSecValueData as String: object], forItemsMatching: secItemQuery)
+        } else {
+            secItemQuery[kSecValueData as String] = object
+            try SecItem.add(attributes: secItemQuery)
+        }
         #endif
     }
     
     // MARK: Removal
     
-    internal static func removeObject(forKey key: String, options: [String : AnyHashable]) -> SecItem.Result {
+    internal static func removeObject(forKey key: String, options: [String : AnyHashable]) throws {
         guard !key.isEmpty else {
-            ErrorHandler.assertionFailure("Can not set a value with an empty key.")
-            return .error(errSecParam)
+            throw KeychainError.emptyKey
         }
         
         var secItemQuery = options
         secItemQuery[kSecAttrAccount as String] = key
         
-        switch SecItem.deleteItems(matching: secItemQuery) {
-        case .success:
-            return .success
-            
-        case let .error(status):
-            switch status {
-            case errSecInteractionNotAllowed, errSecMissingEntitlement:
-                return .error(status)
-                
-            default:
-                // We succeeded as long as we can confirm that the item is not in the keychain.
-                return .success
-            }
-        }
+        try SecItem.deleteItems(matching: secItemQuery)
     }
     
-    internal static func removeAllObjects(matching options: [String : AnyHashable]) -> SecItem.Result {
-        switch SecItem.deleteItems(matching: options) {
-        case .success:
-            return .success
-            
-        case let .error(status):
-            switch status {
-            case errSecInteractionNotAllowed, errSecMissingEntitlement:
-                return .error(status)
-                
-            default:
-                // We succeeded as long as we can confirm that the item is not in the keychain.
-                return .success
-            }
-        }
+    internal static func removeAllObjects(matching options: [String : AnyHashable]) throws {
+        try SecItem.deleteItems(matching: options)
     }
     
     // MARK: Contains
     
-    internal static func containsObject(forKey key: String, options: [String : AnyHashable]) -> SecItem.Result {
+    internal static func containsObject(forKey key: String, options: [String : AnyHashable]) -> OSStatus {
```

Comment:
Nevermind, I saw below your reply to @NickEntin and also I saw how you're using this and I think it makes sense to keep this not throwing. Maybe we do want to change the method name though since `containsObject` sounds like something that will return a Bool?

Reply:
Fair that the name doesn't match the return type. Thoughts on:
```suggestion
    internal static func performCopy(matchingKey key: String, options: [String : AnyHashable]) -> OSStatus {
```

I'm honestly struggling with this one. What this method does under the hood is perform a `SecItemCopyMatching`, but doesn't retrieve the data. It just needs the `OSStatus` return type from the copy action.

Reply:
Or maybe
```suggestion
    internal static func copyItemStatus(forKey key: String, options: [String : AnyHashable]) -> OSStatus {
```

Reply:
My vote is for `performCopy(matching:)`. The fact is returns a status is implied by the return type.

Reply:
Addressed in fd5feba

Reply:
Works for me! Thanks for addressing.